llpc/util/llpcElfWriter.cpp

```diff
@@ -1118,6 +1193,142 @@ void ElfWriter<Elf>::Reinitialize()
+}
+
+// =====================================================================================================================
+// Fixup a descriptor offset relocation (doff_x_y_t symbol) in the ELF with an actual offset value.
+template<class Elf>
+void FixUpDescriptorOffsetRelocation(
+    ElfWriter<Elf> *pWriter,         // [InOut] Writer to target ELF
+    Context *pContext,               // Pipeline compilation context
+    RelocationEntry relocEntry,      // The relocation entry to fixup
+    bool isGraphicsPipeline)         // Whether we are fixing a compute or graphics pipeline
+{
+    size_t idx = 0;
+    const char* pRelocName = relocEntry.name + 5;
+    uint32_t descSet = std::stoi(pRelocName, &idx);
+    pRelocName += idx + 1;
+    idx = 0;
+    uint32_t binding = std::stoi(pRelocName, &idx);
+    pRelocName += idx + 1;
+    ResourceMappingNodeType type = ResourceMappingNodeType::Unknown;
+    switch (*pRelocName)
+    {
+    case 's':
+        type = ResourceMappingNodeType::DescriptorSampler;
+        break;
+    case 'r':
+        type = ResourceMappingNodeType::DescriptorResource;
+        break;
+    case 'b':
+        type = ResourceMappingNodeType::DescriptorBuffer;
+        break;
+    }
+
+    uint32_t offset = 0;
+
+    if (isGraphicsPipeline)
+    {
+        auto pPipelineInfo = reinterpret_cast<const GraphicsPipelineBuildInfo*>(pContext->GetPipelineBuildInfo());
+        offset = GetResourceOffset(descSet, binding, type, pPipelineInfo->vs.pUserDataNodes, pPipelineInfo->vs.userDataNodeCount);
+        if (offset == InvalidValue)
+        {
+            offset = GetResourceOffset(descSet, binding, type, pPipelineInfo->fs.pUserDataNodes, pPipelineInfo->fs.userDataNodeCount);
+        }
+    }
+    else
+    {
+        auto pPipelineInfo = reinterpret_cast<const ComputePipelineBuildInfo*>(pContext->GetPipelineBuildInfo());
+        offset = GetResourceOffset(descSet, binding, type, pPipelineInfo->cs.pUserDataNodes, pPipelineInfo->cs.userDataNodeCount);
+    }
+
+    WriteRelocationConstant(pWriter, relocEntry.offset, offset);
+}
+
+// =====================================================================================================================
+// Fixup relocations in the ELF with actual values.
+template<class Elf>
+void FixUpRelocations(
+    ElfWriter<Elf> *pWriter,                               // [InOut] Writer to target ELF
+    const std::vector<RelocationEntry>& relocations,       // Relocation entries
+    Context *pContext,                                     // Pipeline compilation context
+    bool isGraphicsPipeline)                               // Whether we are fixing a compute or graphics pipeline
+{
+    for(uint32_t i = 0; i < relocations.size(); ++i)
+    {
+        auto reloc = relocations[i];
+        if (strncmp(reloc.name, "doff_", 5) == 0)
+        {
+            FixUpDescriptorOffsetRelocation(pWriter, pContext, reloc, isGraphicsPipeline);
+        }
+    }
+}
```

Comment:
This doesn't feel quite right for the design of a relocation kernel. Could you please refactor this so that:
- there is a function GetRelocationSymbolValue which returns the symbol value (e.g., descriptor offset for "doff_*") and a failure indication if the symbol is not found
- WriteRelocationOffset is effectively inlined here

And some defensive programming to check against the relocation type is necessary. Also, please take a look at the definition of relocation types in AMDGPUUsage.rst in LLVM -- the fixup of the relocation is currently incorrect because it doesn't consider the addend.

Reply:
Done.